Add multiple selection box for results database.

facetothefate opened the pull request:
Reviewed by NOBODY (OOPS!).

* Scripts/libraries/resultsdbpy/resultsdbpy/view/static/library/js/components/TimelineComponents.js:
(XScrollableCanvasProvider):
(Timeline.CanvasSeriesComponent):

gabebear commented:
Submitted a patch at https://bugs.webkit.org/show_bug.cgi?id=195061

JonWBedard commented on Tools/ChangeLog:
```diff
@@ -1,3 +1,13 @@
+2021-11-03  Zhifei Fang  <zhifei_fang@apple.com>
+
+        Add multiple selection box for Timeline library.
+
```
We do need a bugzilla here by policy

JonWBedard commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/library/js/components/TimelineComponents.js:
```diff
@@ -138,11 +149,63 @@ function XScrollableCanvasProvider(exporter, ...childrenFunctions) {
+    containerMouseDownEventStream.action(e => {
+        drag = true;
+        initPos = getMousePosInCanvas(e, containerRef.element);
+        //selectionBoxRef.setState({x: initPos.x, y: initPos.y, width: 0, height: 0, show: true});
```
Why is this commented out?

JonWBedard reviewed (commented):
I would like see this deployed to our staging instance. The video demo is super compelling, but I've found edge cases before by playing with the UI

facetothefate committed:
[results.webkit.org]Add selection box for results database.
https://bugs.webkit.org/show_bug.cgi?id=233958

Reviewed by NOBODY (OOPS!).

Allow user can drag a selection box to select multiple dots.
Allow user to use cmd + click to select multiple dots.
Allow user to use shift + click to select a row of dots.

* Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
(TimelineFromEndpoint):
(TimelineFromEndpoint.prototype.update):
(TimelineFromEndpoint.prototype.getTestResultStatus):
(TimelineFromEndpoint.prototype.getTestResultUrl):
(TimelineFromEndpoint.prototype.render):
* Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/tooltip.js:
* Scripts/libraries/resultsdbpy/resultsdbpy/view/static/library/js/components/TimelineComponents.js:
(pointRectCollisionDetect): Detact if a point and a rect have a insertion.
(rectRectCollisionDetect): Detact if two rects have a insertion.
(XScrollableCanvasProvider): Add selection box.
(xScrollStreamRenderFactory):
(Timeline.CanvasSeriesComponent): Add selection box , cmd + click and shift + click.
(prototype.ExpandableSeriesComponent):
(prototype.Timeline.CanvasContainer):

JonWBedard commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/css/tooltip.css:
```diff
@@ -23,31 +23,34 @@
+ .tooltip {
     z-index: var(--topZIndex);
     position: absolute;
-    opacity: 80%;
```
What is the thinking of dropping the opacity here?

facetothefate replied:
This is for changing the tooltip arrow, I made it use transform instead of border. so it will be a real arrow not a invisible square. The invisible part of that square will cover the dot, so the user cannot click on it.

JonWBedard commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/css/tooltip.css:
```diff
@@ -23,31 +23,34 @@
 }
 
 .tooltip.arrow-up {
-     border-color: transparent transparent #cccd transparent;
```
Why are we dropping the border-color?

facetothefate replied:
ditto.

JonWBedard commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -393,6 +393,8 @@ class TimelineFromEndpoint {
         });
     }
     update() {
+        if (this.radarButtonRef)
```
This should be `issueButtonRef` or something that isn't specific to radar.

facetothefate replied:
I think we will need another bugzilla button ref as well, so I make this radar button.

In my design, we will have two different button, one is for radar and one is for bugzilla

JonWBedard replied:
Then I think we need a list of button refs.

Presumably, we'll be adding GitHub Issues at some point in the future.

JonWBedard commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -512,6 +555,111 @@ class TimelineFromEndpoint {
+                    component = 'JavaScriptCore';
+                break;
+            }
+            const radarUrlComponents = {
```
I think we need to create a more generic API for this.

OpenSource clearly needs to generate bugzilla links instead of radars. Also, things like component setting belong in the application layer, not the library, which will complicate things.

First step for generalizing this is probably to create a generic "createBug" function template taking `title`, `description`, and `classification` for sure, maybe `component` too? You will also probably need to provide some kind of filtering callback so the application instantiating the library can specify it's own components.

facetothefate replied:
This is application layer already. this is not part of library

JonWBedard replied:
I guess we actually have three layers in this application:
- JavaScript libraries
- resultsdbpy "library" that is a packaged application configured by the caller
- Cloud hosted specific application
I'm saying that this code setting components based on test suites belong a level up in the "cloud-hosted specific application." Otherwise this would be the first part of resultsdbpy that references a specific suite.

JonWBedard commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/tooltip.js:
```diff
@@ -79,11 +80,11 @@ class _ToolTip {
                         const rightPoint = stateDiff.points.length > 1 && stateDiff.points[1].x > stateDiff.points[0].x ? stateDiff.points[1] : stateDiff.points[0];
 
                         direction = 'left';
-                        let tipX = leftPoint.x - 12 - bounds.width;
```
This change seems unrelated to the rest of the PR.

facetothefate replied:
This is also for the new tooltip arrow

JonWBedard reviewed (changes requested):
Seems like `TimelineComponents.js` is ready to go, the bits about actually generating the radar need some work though, because we need to support bugzilla as well.
